### DISCUSSION TITLE
Changes for v1.0.0 after all these years

### DISCUSSION THREAD

**Discussion by Samreay** [Samreay/ChainConsumer#100] [Ideas]
@renecotyfanboy here are some ideas I have:

1. Obviously type hint everything
2. Have a single way to specify chain attributes. Right now there's both `c.configure` and `c.add_chain`. I'd like this to become a more explicit:

```python
chain = Chain(df, **options)
consumer.add(chain)
```

So there's only one place to go to.

3. Swap from sphinx to material mkdocs

And a bunch of other smaller things.

What changes were you wanting to make to help with the integration into your tool?

**Comment by renecotyfanboy** [comment]
Hi ! 

> What changes were you wanting to make to help with the integration into your tool?

My first concern was about the maintaining of ChainConsumer in the next years, so glad to see you back. Following your propositions, I deeply agree on a common interface create and use Chain objects. Without being specific to dataframe, I think a class with several constructors would be good depending on the use cases of the community, i.e. 

 ~~~
chain_1 = Chain.from_pandas(df, **kwargs)
chain_2 = Chain.from_dict(dictionary, **kwargs)
chain_3 = Chain.from_arviz(inferenceData, **kwargs)
...
 ~~~

I planned to do this on my tool's side at first but since you're back it would be great to put this directly here 

As I was mentioning in mails, the InferenceData standard is growing in Bayesian computing community, and [arviz](https://python.arviz.org/en/stable/index.html) is really a great tool for diagnosing results, but unusable when it comes to publication, because oh God it is ugly. For my tool, I would have thought of a workflow ending with arviz diagnostics and then results exportation using ChainConsumer, both relying on InferenceData. 

I would be glad to implement here and there some high-level methods to interface this, and in general to help on the maintaining of ChainConsumer following your vision.

**Reply by Samreay** [reply]
From my own experience, I believe the analytics package is mostly unused in chainconsumer. Would you agree? Im trying to see if there's scope for making the package a bit smaller or if I should keep all the current functionality. Anything you think should be dropped?

**Reply by renecotyfanboy** [reply]
In my personal use, I see ChainConsumer as the transition between analysis of chains and actual paper. I do most of my diagnostics using arviz, and export the actual figures and tables for papers using CC since it is pretty and customizable. But still this is my experience, for instance, I've never used the model comparison package in CC. In my opinion, it would be cool to keep the functionalities, and maybe rely on arviz or other packages to shrink the size of CC + get better summary statistics if this seems necessary. WDYT?

**Reply by Samreay** [reply]
Having not used arviz before, I have no thoughts, but a quick google shows me it can do a ton of fun things!

**Reply by Samreay** [reply]
@renecotyfanboy feel free to give 1.0.0-rc1 a whirl, it should be on pypi now, and updated documentation with mkdocs here: https://samreay.github.io/ChainConsumer/

**Reply by renecotyfanboy** [reply]
Great! I'll start to test it today, I am impressed by the amount of work you've done in only 5 days 

And btw thank you for introducing me to mkdocs. It is so awesome!

**Reply by renecotyfanboy** [reply]
I cloned it and ran the tests without any issues ✅ , I'll start working on CC in the coming days!

**Reply by Samreay** [reply]
Awesome! So regarding a `translator` or similar package, there are a few ways I can see this going.

1. We create a translator package and literally provide functions like `arviz_to_dataframe` to make it easier for people to use different data formats. The syntax would then be something like:

```python
data = some_arviz_thing
df = arviz_to_dataframe(data)
Chain(samples=df, name="Blah")
```

The downside is that the arviz might specify the name, not just the dataframe.

We could try and get around this by having a function call instead called:

`arviz_to_chain`, but then the downside is, if you want to specify Chain kwargs, you'd lose the type hinting, because the function would look like:

```python
def arviz_to_chain(data: whatever_type, **kwargs):
    df, name = blahblahblah
    return Chain(samples=df, name=name, **kwargs)
```

And I'm really trying to *not* lose that kwarg info because it leads to a lot of questions about "How do I do X" when it should be obvious from the autocompletion.

Another approach would be for Chain to take more than a dataframe as samples, and we swap from accessing `samples` under the hood, to accessing a cached and computed parsed version, like so

```python
translators = {
    pd.DataFrame: lambda x: x,
    arviz.thing: somefunc
}

class Chain(BetterBase):
    samples: pd.DataFrame | arviz.thing = Field(default=..., description="Blah", frozen=True)
    ...
    
    @cached_property
    def final_samples(self) -> pd.DataFrame:
        the_type = type(self.samples)
        assert the_type in translators, f"Type {the_type} does not have a translation function available"
        return translators[the_type](self.samples)
```

That I think would make the most sense for a user, and also be easy to work with in the back end. Additionally, in the `model_validator` we could do things like extract the name or other properties (like number of effective data points, number free parameters, weights, posterior, etc) based on the type of `samples` and what's available.
    
    
 The main thing to be careful of with this approach is not requiring all those libraries by default and just doing conditional imports and using stringed type hints

**Reply by renecotyfanboy** [reply]
My take on this would have been a custom constructor for chain : 

``` python
class Chain(BetterBase):
    samples: pd.DataFrame #and nothing else!

    def __init__(self, samples: pd.Dataframe = None, name: str = "Name", **kwargs):
        ...

    @classmethod
    def from_arviz_id(cls, inference_data: az.InferenceData,  name: str = "Name", **kwargs):

        # We implemente the converter here
        ====
        converted_df = translator(inference_data) 
        ====

        return cls(samples = converted_df, name=name, **kwargs)

my_favorite_chain = Chain.from_arviz_id(inference_data, name="I <3 ID")
```

In my opinion, this kind of pattern feels more natural since it allows for extension (i.e. making a constructor for chain from a dict instance or anything else would simply add a new constructor for Chain instead of storing a list of translator functions in a dictionary). It also keeps a simple internal typing for the Chain class. This leads to pretty self-explanatory code in addition to typing. 

I think building external translator functions would be a bit overkill, and that we should simply store a DF in each chain, as you mentioned. This does not solve the issue of strong dependency on arviz to declare the typing of variable tho. WDYT ?

**Reply by Samreay** [reply]
Yeah, agreed that having it all in the `Chain` class makes it easier to use. Alright so two options then:

#### Option One: Class methods to handle each type. `Chain.from_arviz`, `Chain.from_numpro`, etc

Benefits: Using class constructors is very common and easy to understand.
Downsides: Kwarg autocompletion is harder to infer due to the `**kwarg` unpack

#### Option Two: `field_validator` allows arbitrary types as an input and we translate then based upon the type.

Benefits: Single constructor and no issues with kwarg passthrough
Downsides: Assumes that all conversion can be done without any additional arguments, which reduces extensibility.

I think the downsides in option two make option one (the one you agreed was nicer) the better choice. Do you have some sample code to convert from az.InferenceData to a pandas dataframe (along with a sample datafile) you could throw my way, and then I can also play around in code rather than comments?

**Reply by renecotyfanboy** [reply]
Yes of course ! 

``` python
import numpy as np
import arviz as az

inference_data = az.from_json('fit.json')
dataframe = inference_data.to_dataframe(groups= "posterior").drop(columns=["chain", "draw"])
```

This is pretty straightforward, and as I am writing this, I am thinking that it is pretty much an overkill to integrate this in CC since it is that easy. When thinking of this feature, I had in mind edge cases where some variables are multidimensional (i.e. NN layers), but well, I forgot that these are cases where CC wouldn't be adapted and used… 

Still, I joined a .json file containing chains from a fit if you want to play with it

[fit.json](https://github.com/Samreay/ChainConsumer/files/12903104/fit.json)

**Reply by Samreay** [reply]
Ah yeah that is fairly simple. Hmm, do you know of any other slightly more complex data formats that would be good to be supported. numpyro?

**Reply by renecotyfanboy** [reply]
numpyro's results are also simple to gather too, the following code should do the trick 

```python
import pandas as pd

# using JAX (should be available in a numpyro env)
from jax.tree_util import tree_map 
mcmc: numpyro.infer.MCMC
samples = tree_map(np.flatten, mcmc.get_samples())
dataframe = pd.DataFrame.from_dict(samples)

# no JAX should work too
samples = {key:np.flatten(value) for key, value in mcmc.get_samples().items()}
dataframe = pd.DataFrame.from_dict(samples)
```

You might take a lookf at [arviz's InferenceData constructors](https://python.arviz.org/en/stable/getting_started/CreatingInferenceData.html) which shows the currently used bayesian libraries in python. It might be a little outdated since there is still pymc3 in it. On my side I've only used emcee, pymc3-5 and pyro and numpyro. I think emcee is by far the most popular library in astro-ph due to its simplicity, and I've seen an issue (#76) related to their hdf5 storing format. From what I read, it [can be loaded](https://emcee.readthedocs.io/en/stable/user/backends/) easily at the cost of additional hdf5 dependency. Also, it is simple to extract samples directly from the sampler in memory using `sampler.get_chain()`.

**Reply by Samreay** [reply]
Ah, its difficult to code things up when its been actual years since I've used anything at all from this. But okay, I could use jax to provide a simple example for numpyro that anyone can use to extend to other formats, like the hdf5 format

**Reply by renecotyfanboy** [reply]
I can take care of this if you want, specify me what you have in mind, and I'll push in on a new branch with pleasure!
I edited the previous post with a no JAX solution too

**Reply by Samreay** [reply]
Oh sure thing. So what I was going to do was:

1. Add jax or whatever is needed using poetry as an `--extra` dependency
2. Create a translator package and put a `from_jax.py` file in it, which would have a single function and no global jax import
3. Do the normal try and import, raise an error if jax isnt installed
4. Create a thin classmethod like you have in your comment that simply calls the translator, gathers the dataframe and any other kwargs back (maybe some formats specify a chain name, or have a specific column they put the weights in), and then invokes chain with the combination of the translation package plus any other kwargs
5. Add a test or two using some dummy data

**Reply by snwnde** [reply]
> #### Option One: Class methods to handle each type. `Chain.from_arviz`, `Chain.from_numpro`, etc
> 
> Benefits: Using class constructors is very common and easy to understand.
> Downsides: Kwarg autocompletion is harder to infer due to the `**kwarg` unpack

For the record, the downsides are alleviated upon the adoption of [PEP 692](https://peps.python.org/pep-0692/) and [pylance already supports it](https://github.com/microsoft/pylance-release/issues/2541#issuecomment-1448576066).

**Reply by Samreay** [reply]
@snwnde Yeah thats a big step forward, I just dont want to have to redefine everything as a TypeDict. I raised a feature request over on the python typing repo to see if they'd be open to converting providing Unpack compatibility with a dataclass, or a way to easily go from dataclass to typeddict (without needing code to be run).

Feel free to comment and upvote it here: https://github.com/python/typing/issues/1495

**Comment by Samreay** [comment]
Current changes are in  #101

What a weekend its been

**Comment by Samreay** [comment]
@renecotyfanboy 1.0.0 going out now!

**Reply by renecotyfanboy** [reply]
It appears that the PlotConfig object is not documented, do you plan to make a specific section or just include it through its docstring ?

**Reply by Samreay** [reply]
Just forgot the line, should be out now
